### PR TITLE
Add nodejs version check

### PR DESCRIPTION
**Pull request by huozhi**
There're some reports related to using incorrect Node.js version, sometimes users ended up in weird errors. As pkg manager like pnpm only gives a warning ` WARN  Unsupported engine: wanted: {"node":">=18.17.0"} (current: {"node":"v16.18.0","pnpm":"8.9.0"})` which is not easy to investigate the issue.

We're adding a version check at the beginning of the Next.js process so if Node.js version is mis-match the version specified in package.json, bail with error.

Examples

node 16.x
```
You are using Node.js 16.18.0. Node.js >= v18.17.0 is required.
```

node.18.16
```
You are using Node.js 18.16.1. Node.js >= v18.17.0 is required.
```

### CODE REVIEW

**Review by timneutkens — DISMISSED** [review]

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 443d654db110c4dab4cacf441ebd0bd4b21d88fa

`pnpm test-dev test/e2e/app-dir/app-compilation/index.test.ts` 
- app dir > HMR > should not cause error when removing loading.js

<details>
<summary>Expand output</summary>

  ● app dir › HMR › should not cause error when removing loading.js

    TIMED OUT: hello from new page

    hello from slow page

    undefined

      626 |
      627 |   if (hardError) {
    > 628 |     throw new Error('TIMED OUT: ' + regex + '\n\n' + content + '\n\n' + lastErr)
          |           ^
      629 |   }
      630 |   return false
      631 | }

      at check (lib/next-test-utils.ts:628:11)
      at Object.<anonymous> (e2e/app-dir/app-compilation/index.test.ts:44:11)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by abenhamdine** [comment]
By doing this, you add some duplicated logic with package.json `engines ` property, wouldn't it be better to read package.json to have a single place where this logic resides ?

File: packages/next/src/bin/next.ts [hunk 1]
```diff
@@ -115,6 +115,18 @@ async function main() {
...
   }
 
+  const nodeVersions = process.versions.node.split('.').map((v) => Number(v))
+  // Node.js version check, if it's less than v18 or v18.17, quit the process.
+  if (
+    nodeVersions[0] < 18 ||
+    (nodeVersions[0] === 18 && nodeVersions[1] < 17)
+  ) {
+    console.error(
+      `You are using Node.js ${process.versions.node}. Node.js >= v18.17 is required.`
```
Comment (pvdz) [inline packages/next/src/bin/next.ts:125]:
```suggestion
      `You are using Node.js ${process.versions.node}. For NextJS, Node.js >= v18.17 is required.`
```

**Review by pvdz — COMMENTED** [review]
Nit: Can we add the source of this message to state explicitly being from nextjs? Otherwise it's almost as cryptic as the package manager itself. (It also works around that awkward "Node.js vx. Node.js vy" back to back)

**Comment by huozhi** [comment]
@pvdz thanks for checking, update the error message!
@abenhamdine good suggestion, I've updated the PR to pre-compile the required nodejs version into process.env and check it here

**Review by pvdz — APPROVED** [review]

**Review by danielbarion — APPROVED** [review]
A simple change but helpful while debugging node error-related, thanks!

**Review by ijjk — DISMISSED** [review]

**Review by ztanner — APPROVED** [review]

**Review by acdlite — APPROVED** [review]

**Pull request merged**